src/qibo/models.py

```diff
@@ -212,12 +225,15 @@ def _loss(params, circuit, hamiltonian):
             loss = _loss
         else:
             loss = lambda p, c, h: _loss(p, c, h).numpy()
-        result, parameters = self.optimizers.optimize(loss, initial_state,
-                                                      args=(self.circuit, self.hamiltonian),
-                                                      method=method, options=options,
-                                                      compile=compile, processes=processes)
+        result, parameters, extra = self.optimizers.optimize(loss, initial_state,
+                                                             args=(self.circuit, self.hamiltonian),
+                                                             method=method, options=options,
+                                                             compile=compile, processes=processes,
+                                                             jac=jac, hess=hess, hessp=hessp,
+                                                             bounds=bounds, constraints=constraints,
+                                                             tol=tol, callback=callback)
         self.circuit.set_parameters(parameters)
-        return result, parameters
+        return result, parameters, extra
```

Comment:
We could make the optimizers and the `minimize` methods of models return a single result object instead of three objects. For scipy and the cma optimizer this can be the `extra` object since it already contains `result` and `parameters`, while for SGD we could create a small custom object that follows the scipy structure, eg. the result and parameters are accessible via `result.fun` and `result.x`. This would change our existing API slightly but I think it is easier to maintain. Otherwise the current structure is also fine with me. Let me know what you think.

Reply:
Concerning this point, are you suggesting to define a new return object for all optimizer?
If we return the extra object for scipy and cma they will have different attributes (.fun, .x vs .result.fbest, .result.xbest).

Reply:
Yes, the suggestion was to return a single object for all optimizers which could follow the scipy conventions. For cma, indeed the attributes differ from scipy so we could either return our own custom object that follows scipy or return the cma object but break consistency with either optimizers. I am not sure if returning a single object is simpler than what we have now from the user perspective so we can keep whatever you prefer (or ask people who use the optimizers for their preference)

Reply:
Right, I think we can keep the layout as it is for now and then see if there are more specific request in the future.